File: docs/compat/sentry-python/senden.py
"""Beispiel: sentry-python

Schickt Nachricht, Fehler, Transaktion und Sitzung mit dem offiziellen
Python-SDK. Getauscht wird nur die DSN:

    python -m venv .venv
    .venv/Scripts/pip install -r requirements.txt     # Linux/macOS: .venv/bin/pip

    # gegen den laufenden Klon
    SENTRY_DSN="http://<public_key>@localhost:8000/1" .venv/Scripts/python senden.py

    # gegen den Mitschnitt-Server (Aufnahme neu erstellen)
    .venv/Scripts/python senden.py
"""

import os
import time

import sentry_sdk

dsn = os.environ.get(
    "SENTRY_DSN", "http://aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa@127.0.0.1:9911/1"
)

sentry_sdk.init(
    dsn=dsn,
    release="compat@1.0.0",
    environment="compat",
    # Ohne das schickt das SDK die Transaktion nicht ab, sondern verwirft sie
    # als nicht gezogene Stichprobe.
    traces_sample_rate=1.0,
    attach_stacktrace=True,
    server_name="compat-beispiel",
)

sentry_sdk.set_user({"id": "4711", "username": "kompatibilitaet"})
sentry_sdk.set_tag("beispiel", "sentry-python")
sentry_sdk.add_breadcrumb(
    category="beispiel", message="Rechnung geladen", level="info"
)

# 0. Eine Sitzung von Hand — sie umschließt alles Weitere. Nicht über
#    `auto_session_tracking`: dessen Sitzung schickt das SDK zusammengefasst und
#    zeitgesteuert ab, ein Beispielskript ist vorher fertig.
sentry_sdk.start_session(session_mode="application")

# 1. Eine Nachricht ohne Fehler.
sentry_sdk.capture_message(
    "Kompatibilitätsprobe: Nachricht aus sentry-python", level="info"
)
sentry_sdk.flush()

# 2. Ein echter Fehler mit Stacktrace und Ursache.
try:
    try:
        raise ValueError("Rechnungsnummer 4711 ist unbekannt")
    except ValueError as ursache:
        raise RuntimeError("Rechnung konnte nicht erstellt werden") from ursache
except RuntimeError as fehler:
    sentry_sdk.capture_exception(fehler)

sentry_sdk.flush()

# 3. Eine Transaktion mit zwei Einzelschritten.
with sentry_sdk.start_transaction(op="http.server", name="GET /rechnungen"):
    for op, beschreibung in (
        ("db.sql.query", "select * from invoices"),
        ("http.client", "GET https://zahlungen.example/status"),
    ):
        with sentry_sdk.start_span(op=op, description=beschreibung):
            time.sleep(0.015)

sentry_sdk.flush()

# 4. Die Sitzung beenden. Abgeschickt wird sie erst beim Schließen des Clients:
#    Sitzungen sammelt das SDK und gibt sie gebündelt ab.
sentry_sdk.end_session()
sentry_sdk.get_client().close()

print(
    f"sentry-python: Nachricht, Fehler, Transaktion und Sitzung an {dsn} geschickt"
)

File: docs/compat/sentry-node/senden.mjs
/*
|--------------------------------------------------------------------------
| Beispiel: @sentry/node
|--------------------------------------------------------------------------
|
| Schickt Nachricht, Fehler, Transaktion und Sitzung mit dem offiziellen
| Node-SDK. Getauscht wird nur die DSN:
|
|     npm install
|
|     # gegen den laufenden Klon
|     SENTRY_DSN="http://<public_key>@localhost:8000/1" npm run senden
|
|     # gegen den Mitschnitt-Server (Aufnahme neu erstellen)
|     npm run senden
|
| Die Sitzung kommt aus diesem Beispiel und nicht aus dem PHP-Beispiel: Release
| Health gibt es im PHP-SDK nicht.
*/

import * as Sentry from '@sentry/node';

const dsn = process.env.SENTRY_DSN || 'http://aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa@127.0.0.1:9911/1';

Sentry.init({
    dsn,
    release: 'compat@1.0.0',
    environment: 'compat',
    // Ohne das schickt das SDK die Transaktion nicht ab, sondern verwirft sie
    // als nicht gezogene Stichprobe.
    tracesSampleRate: 1.0,
    // Ein Beispielskript läuft und ist gleich wieder weg; ohne das würde jede
    // Meldung einzeln gebündelt und die Reihenfolge im Mitschnitt wäre Zufall.
    maxBreadcrumbs: 10,
});

Sentry.setUser({ id: '4711', username: 'kompatibilitaet' });
Sentry.setTag('beispiel', 'sentry-node');
Sentry.addBreadcrumb({ category: 'beispiel', message: 'Rechnung geladen', level: 'info' });

// 1. Eine Sitzung — sie umschließt alles Weitere, so wie eine Anwendung sie um
//    ihre Laufzeit legt.
Sentry.startSession();

// 2. Eine Nachricht ohne Fehler.
Sentry.captureMessage('Kompatibilitätsprobe: Nachricht aus @sentry/node', 'info');
await Sentry.flush(5000);

// 3. Ein echter Fehler mit Stacktrace und Ursache.
try {
    try {
        throw new TypeError('Rechnungsnummer 4711 ist unbekannt');
    } catch (ursache) {
        throw new Error('Rechnung konnte nicht erstellt werden', { cause: ursache });
    }
} catch (fehler) {
    Sentry.captureException(fehler);
}
await Sentry.flush(5000);

// 4. Eine Transaktion mit zwei Einzelschritten.
await Sentry.startSpan({ name: 'GET /rechnungen', op: 'http.server' }, async () => {
    for (const [op, beschreibung] of [
        ['db.sql.query', 'select * from invoices'],
        ['http.client', 'GET https://zahlungen.example/status'],
    ]) {
        await Sentry.startSpan({ op, name: beschreibung }, async () => {
            await new Promise((fertig) => setTimeout(fertig, 15));
        });
    }
});
await Sentry.flush(5000);

// 5. Die Sitzung beenden — das schickt sie ab.
Sentry.endSession();
await Sentry.flush(5000);

console.log(`@sentry/node: Nachricht, Fehler, Transaktion und Sitzung an ${dsn} geschickt`);
